packages/marshal/src/helpers/passStyle-helpers.js

```diff
@@ -102,36 +105,64 @@ harden(checkNormalProperty);
 harden(getTag);
 
-/**
- * @param {{ [PASS_STYLE]: string }} tagRecord
- * @param {PassStyle} passStyle
- * @param {Checker} [check]
- * @returns {boolean}
- */
-export const checkTagRecord = (tagRecord, passStyle, check) => {
-  const reject = !!check && (details => check(false, details));
+export const checkPassStyle = (obj, expectedPassStyle, check) => {
+  const actual = obj[PASS_STYLE];
   return (
-    (isObject(tagRecord) ||
-      (reject &&
-        reject(X`A non-object cannot be a tagRecord: ${tagRecord}`))) &&
-    (isFrozen(tagRecord) ||
-      (reject && reject(X`A tagRecord must be frozen: ${tagRecord}`))) &&
-    (!isArray(tagRecord) ||
-      (reject && reject(X`An array cannot be a tagRecords: ${tagRecord}`))) &&
-    checkNormalProperty(tagRecord, PASS_STYLE, false, check) &&
-    (tagRecord[PASS_STYLE] === passStyle ||
-      (reject &&
-        reject(
-          X`Expected ${q(passStyle)}, not ${q(
-            tagRecord[PASS_STYLE],
-          )}: ${tagRecord}`,
-        ))) &&
-    checkNormalProperty(tagRecord, Symbol.toStringTag, false, check) &&
-    (typeof getTag(tagRecord) === 'string' ||
-      (reject &&
-        reject(
-          X`A [Symbol.toStringTag]-named property must be a string: ${tagRecord}`,
-        )))
+    actual === expectedPassStyle ||
+    (!!check &&
+      check(
+        false,
+        X`Expected ${q(expectedPassStyle)}, not ${q(actual)}: ${obj}`,
+      ))
   );
 };
-harden(checkTagRecord);
+harden(checkPassStyle);
+
+const makeCheckTagRecord = checkProto => {
+  /**
+   * @param {{ [PASS_STYLE]: string }} tagRecord
+   * @param {PassStyle} passStyle
+   * @param {Checker} [check]
+   * @returns {boolean}
+   */
+  const checkTagRecord = (tagRecord, passStyle, check) => {
+    const reject = !!check && (details => check(false, details));
+    return (
+      (isObject(tagRecord) ||
+        (reject &&
+          reject(X`A non-object cannot be a tagRecord: ${tagRecord}`))) &&
+      (isFrozen(tagRecord) ||
+        (reject && reject(X`A tagRecord must be frozen: ${tagRecord}`))) &&
+      (!isArray(tagRecord) ||
+        (reject && reject(X`An array cannot be a tagRecord: ${tagRecord}`))) &&
+      checkNormalProperty(tagRecord, PASS_STYLE, false, check) &&
+      checkPassStyle(tagRecord, passStyle, check) &&
+      checkNormalProperty(tagRecord, Symbol.toStringTag, false, check) &&
+      (typeof getTag(tagRecord) === 'string' ||
+        (reject &&
+          reject(
+            X`A [Symbol.toStringTag]-named property must be a string: ${tagRecord}`,
+          ))) &&
+      checkProto(tagRecord, getPrototypeOf(tagRecord), check)
+    );
+  };
+  return harden(checkTagRecord);
+};
+
+export const checkTagRecord = makeCheckTagRecord(
+  (val, proto, check) =>
+    proto === objectPrototype ||
+    (!!check &&
+      check(false, X`A tagRecord must inherit from Object.prototype: ${val}`)),
+);
+
+export const checkFunctionTagRecord = makeCheckTagRecord(
```

Comment:
What is our stance on builtin `Function` subclasses, such as async functions, generator functions, or async generators? Can these be far functions? These do not *directly* inherit from `Function.prototype`, but they do inherit from known primordial prototypes.

Reply:
This preserves the existing behavior, which allows far async functions if their prototype chain has been suitably manipulated (e.g., `setPrototypeOf(async () => {}, makeTagishRecord('Remotable', getPrototypeOf(async () => {})))`) while rejecting generators and async generators—and classes, for that matter—by dint of a non-removable "prototype" property ("_Far functions unexpected properties besides .name and .length_"). But an (async)generator is basically a callable factory for (async)iterator objects which at their core are a bag of inherited methods, so I could definitely see extending support in that direction, probably with a helper like `ExoGenerator(async function* () { … })` to alter the prototype chain and harden responses.